src/sgx-verifier/AttestationVerifier.sol

```diff
@@ -0,0 +1,207 @@
+     ***********************/
+
+    /// @notice The address of automata's DCAP Attestation contract.
+    IDcapAttestation public immutable attestationVerifier;
+
+    /*********************
+     * Storage Variables *
+     *********************/
+
+    /// @dev The list of trusted enclaves.
+    EnumerableSet.Bytes32Set private trustedUserMrEnclave;
+
+    /// @dev The list of trusted signers.
+    EnumerableSet.Bytes32Set private trustedUserMrSigner;
+
+    /***************
+     * Constructor *
+     ***************/
+
+    constructor(address _attestationVerifierAddr) {
+        attestationVerifier = IDcapAttestation(_attestationVerifierAddr);
+    }
+
+    /*************************
+     * Public View Functions *
+     *************************/
+
+    /// @inheritdoc IAttestationVerifier
+    function isTrustedMrSigner(bytes32 mrSigner) public view returns (bool) {
+        return trustedUserMrSigner.contains(mrSigner);
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function isTrustedMrEnclave(bytes32 mrEnclave) public view returns (bool) {
+        return trustedUserMrEnclave.contains(mrEnclave);
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function getTrustedMrSigners() external view returns (bytes32[] memory signers) {
+        uint256 length = trustedUserMrSigner.length();
+        signers = new bytes32[](length);
+        for (uint256 i = 0; i < length; i++) {
+            signers[i] = trustedUserMrSigner.at(i);
+        }
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function getTrustedMrEnclaves() external view returns (bytes32[] memory enclaves) {
+        uint256 length = trustedUserMrEnclave.length();
+        enclaves = new bytes32[](length);
+        for (uint256 i = 0; i < length; i++) {
+            enclaves[i] = trustedUserMrEnclave.at(i);
+        }
+    }
+
+    /// @inheritdoc IAttestationVerifier
+    function verifyAttestation(bytes calldata _report, bytes32 _userData) external view {
+        (bool success, bytes memory output) = attestationVerifier.verifyAndAttestOnChain(_report);
+        if (!success) {
+            revert ErrorInvalidReport();
+        }
+
+        (bytes32 reportUserData, bytes32 mrEnclave, bytes32 mrSigner) = extractEnclaveReport(output);
+        if (reportUserData != _userData) {
+            revert ErrorReportDataMismatch();
+        }
+        // check local enclave report
+        if (!isTrustedMrEnclave(mrEnclave)) {
+            revert ErrorInvalidMrEnclave();
+        }
+        if (!isTrustedMrSigner(mrSigner)) {
+            revert ErrorInvalidMrSigner();
+        }
+    }
+
+    /************************
+     * Restricted Functions *
+     ************************/
+
+    /// @notice Update the status of a signer.
+    /// @param _mrSigner The signer to update.
+    /// @param _status The new status to update. If it is `true`, the signer is trusted.
+    function updateMrSigner(bytes32 _mrSigner, bool _status) external onlyOwner {
```

Comment:
What is exactly `mrSigner`? How and when will we call this?

Reply:
`mrSigner` is the one who compiles the enclave problem and sign the program hash.

Reply:
Got it. Since we already sign the transaction when we update `mrEnclave` (the program hash), what is the use of `mrSigner`?

My understanding was this signing entity would allow us to update the program hash just by verifying the signature. Useful when the program changes frequently. But it seems we don't need this functionality.

Reply:
`mrSigner` and `mrEnclave` is part of the attestation report, we can recover them from the report and check whether the recovered one is whitelisted. Just some cross validation to make sure everything is ok.